### PR TITLE
Add volume control

### PR DESCRIPTION
**Pull request by Diegovsky**
Currently it works via the mpris interface but allows implementing via a GUI slider futurely.

I added some new events and an action as I was groking the codebase so I'm not sure if I edited the wrong places

### CODE REVIEW

File: src/dbus/types.rs [hunk 1]
```diff
@@ -167,9 +168,23 @@ impl SharedMprisState {
...
     }
 
+    pub fn volume(&self) -> f64 {
+        self.0
+            .lock()
+            .map(|s| s.volume)
+            .map_err(|e| error!("Failed to get volume: {:?}", e))
+            .unwrap_or(1.0f64)
+    }
+    pub fn set_volume(&self, volume: f64) {
```
Comment (xou816) [inline src/dbus/types.rs:182]: (nitpick) could use a space above the function

File: src/dbus/mpris.rs [hunk 2]
```diff
@@ -314,9 +314,14 @@ impl SpotMprisPlayer {
...
     #[dbus_interface(property)]
     pub fn volume(&self) -> f64 {
-        0f64
+        self.state.volume()
     }
 
     #[dbus_interface(property)]
-    pub fn set_volume(&self, value: f64) {}
+    pub fn set_volume(&self, value: f64) -> Result<()> {
+        self.state.set_volume(value);
```
Comment (xou816) [inline src/dbus/mpris.rs:322]: right now its not a big deal at all, but in the long term I´d recommend not modifying the state here, but only when the change has been acknowledged by the main process (written to the app state). this keeps all mutations of the state (and dbus outbound notifications) in one place, whether this happened through mpris or through something else (like a slider) :)

File: src/dbus/mpris.rs [hunk 3]
```diff
@@ -9,7 +9,7 @@ use zbus::ObjectServer;
 use zvariant::ObjectPath;
 
 use super::types::*;
-use crate::app::state::RepeatMode;
+use crate::app::state::{PlaybackEvent, RepeatMode};
```
Comment (xou816) [inline src/dbus/mpris.rs:12]: unused import apparently

File: src/player/player.rs [hunk 4]
```diff
@@ -6,10 +6,12 @@ use librespot::core::config::SessionConfig;
 use librespot::core::keymaster;
 use librespot::core::session::{Session, SessionError};
 
+use librespot::playback::mixer::softmixer::SoftMixer;
+use librespot::playback::mixer::{AudioFilter, Mixer};
```
Comment (xou816) [inline src/player/player.rs:10]: AudioFilter is not used either apparently

**Review by xou816 — COMMENTED** [review]
Hey! Thanks a lot for this, I had no idea how this would be implemented on the librespot side, nice! This is a really helpful feature. Looks good to me overall! Just a few minor comments

File: src/dbus/mpris.rs (hunk 2 above)
Comment (Diegovsky) [reply src/dbus/mpris.rs:322]: I agree! How would I get the volume from the main process from mpris?

File: src/dbus/mpris.rs (hunk 3 above)
Comment (Diegovsky) [reply src/dbus/mpris.rs:12]: woops must be because of my refactor

**Comment by Diegovsky** [comment]
> Hey! Thanks a lot for this, I had no idea how this would be implemented on the librespot side, nice!

Yeah they don't have docs so I had to be a little creative lol

I'm honestly impressed I could pull this off :)

File: src/player/player.rs (hunk 4 above)
Comment (Diegovsky) [reply src/player/player.rs:10]: I'm pretty sure it's because of the trait but I'll fix it in the next commit

File: src/dbus/types.rs (hunk 1 above)
Comment (Diegovsky) [reply src/dbus/types.rs:182]: noted

File: src/dbus/mpris.rs (hunk 2 above)
Comment (xou816) [reply src/dbus/mpris.rs:322]: my files are really poorly organized but you can see that in the `impl EventListener for AppPlaybackStateListener`

Comment (Diegovsky) [reply src/dbus/mpris.rs:322]: Nice! I'll be comitting my changes soon but I'll take some days to respond since it's new years

**Comment by Diegovsky** [comment]
I just uploaded the new changes and also fixed a problem with librespot's SoftMixer; apparently it starts with 50% volume instead of 100% so I had to hard code it to 100% to be consistent with initial values.

**Comment by xou816** [comment]
Perfect, good thinking with the clamping too!!

Thanks a lot, merging this!

**Pull request merged**